zboxcore/sdk/downloadworker.go

```diff
@@ -46,338 +45,494 @@ type DownloadRequest struct {
+	)
 
-	startBlock := req.startBlock
-	endBlock := req.endBlock
-	numBlocks := req.numBlocks
+	f, err := req.getFileHandler(remotePathCB)
+	if err != nil {
+		logger.Logger.Error(err)
+		return
+	}
+	defer f.Close()
+
+	var isFullDownload bool
+	fileHasher := createDownloadHasher(req.chunkSize, req.datashards, fRef.EncryptedKey != "")
+	var mW io.Writer
+	if req.startBlock == 0 && req.endBlock == chunksPerShard {
+		isFullDownload = true
+		mW = io.MultiWriter(fileHasher, f)
+	} else {
+		mW = io.MultiWriter(f)
+	}
+
+	err = req.initEC()
+	if err != nil {
+		logger.Logger.Error(err)
+		return
+	}
+	if req.encryptedKey != "" {
+		req.initEncryption()
+	}
+
+	var downloaded int
+	startBlock, endBlock, numBlocks := req.startBlock, req.endBlock, req.numBlocks
+	// remainingSize should be calculated based on startBlock number
+	// otherwise end data will have null bytes.
+	remainingSize := size - startBlock*int64(req.effectiveChunkSize)
+
+	if req.statusCallback != nil {
+		// Started will also initialize progress bar. So without calling this function
+		// other callback's call will panic
+		req.statusCallback.Started(req.allocationID, remotePathCB, OpDownload, int(size))
+	}
 
 	for startBlock < endBlock {
-		cnt := startBlock
-		logger.Logger.Info("Downloading block ", cnt+1)
-		if (startBlock + numBlocks) > endBlock {
+		if startBlock+numBlocks > endBlock {
 			numBlocks = endBlock - startBlock
 		}
+		logger.Logger.Info("Downloading block ", startBlock+1, " - ", startBlock+numBlocks)
 
-		data, err := req.downloadBlock(cnt+1, int(numBlocks))
+		data, err := req.getBlocksData(startBlock+1, numBlocks, req.downloadMask, req.consensusThresh)
 		if err != nil {
-			sys.Files.Remove(req.localpath)
-			if req.statusCallback != nil {
-				req.statusCallback.Error(req.allocationID, remotePathCallback, OpDownload, errors.Wrap(err, fmt.Sprintf("Download failed for block %d. ", cnt+1)))
-			}
+			req.errorCB(errors.Wrap(err, fmt.Sprintf("Download failed for block %d. ", startBlock+1)), remotePathCB)
 			return
 		}
 		if req.isDownloadCanceled {
-			req.isDownloadCanceled = false
-			sys.Files.Remove(req.localpath)
-			if req.statusCallback != nil {
-				req.statusCallback.Error(req.allocationID, remotePathCallback, OpDownload, errors.New("", "Download aborted by user"))
-			}
+			req.errorCB(errors.New("download_abort", "Download aborted by user"), remotePathCB)
 			return
 		}
 
-		n := int64(math.Min(float64(size), float64(len(data))))
+		n := int64(math.Min(float64(remainingSize), float64(len(data))))
 		_, err = mW.Write(data[:n])
 
 		if err != nil {
-			sys.Files.Remove(req.localpath) //nolint: errcheck
-			if req.statusCallback != nil {
-				req.statusCallback.Error(req.allocationID, remotePathCallback, OpDownload, errors.Wrap(err, "Write file failed"))
-			}
+			req.errorCB(errors.Wrap(err, "Write file failed"), remotePathCB)
 			return
 		}
 		downloaded = downloaded + int(n)
-		size = size - n
+		remainingSize -= n
```

Comment:
This check is unnecessary because `startBlock` and `endBlock` are calculated based on size to download.
It will only be used in:
```
n := int64(math.Min(float64(remainingSize), float64(len(
_, err = mW.Write(data[:n])
```

`remainingSize` won't get chance to be negative.